src/btree/bt_page.c

```diff
@@ -15,6 +15,434 @@ static int __inmem_row_int(WT_SESSION_IMPL *, WT_PAGE *, size_t *);
 
+/*
+ * __page_build_ref --
+ *     Create a ref from a base image or a delta.
+ */
+static int
+__page_build_ref(WT_SESSION_IMPL *session, WT_REF *parent_ref, WT_CELL_UNPACK_ADDR *base_key,
+  WT_CELL_UNPACK_ADDR *base_val, WT_CELL_UNPACK_DELTA_INT *delta, bool base, WT_REF **refp,
+  size_t *incrp)
+{
+    WT_REF *ref;
+    uint8_t key_type, value_type;
+
+    WT_ASSERT(session, incrp != NULL);
+
+    WT_RET(__wt_calloc_one(session, refp));
+    *incrp += sizeof(WT_REF);
+
+    ref = *refp;
+    ref->home = parent_ref->page;
+    key_type = base ? base_key->type : delta->key.type;
+    value_type = base ? base_val->type : delta->value.type;
+
+    switch (key_type) {
+    case WT_CELL_KEY:
+        if (base)
+            __wt_ref_key_onpage_set(parent_ref->page, ref, base_key);
+        else {
+            WT_RET(__wti_row_ikey(session, 0, delta->key.data, delta->key.size, ref));
+            *incrp += sizeof(WT_IKEY) + delta->key.size;
+        }
+        break;
+    case WT_CELL_KEY_OVFL:
+    /* Overflow keys are not supported. */
+    default:
+        WT_RET(__wt_illegal_value(session, delta->key.type));
+    }
+
+    switch (value_type) {
+    case WT_CELL_ADDR_INT:
+        F_SET(ref, WT_REF_FLAG_INTERNAL);
+        break;
+    case WT_CELL_ADDR_LEAF:
+    case WT_CELL_ADDR_LEAF_NO:
+        F_SET(ref, WT_REF_FLAG_LEAF);
+        break;
+    case WT_CELL_ADDR_DEL:
+    /* Fast truncated pages are not supported. */
+    default:
+        WT_RET(__wt_illegal_value(session, delta->value.type));
+    }
+
+    if (base)
+        ref->addr = base_val->cell;
+    else
+        WT_RET(__wt_memdup(session, delta->value.data, delta->value.size, &ref->addr));
+    return (0);
+}
+
+/*
+ * __page_merge_internal_delta_with_base_image --
+ *     Merge the consolidated delta array with the base image.
+ */
+static int
+__page_merge_internal_delta_with_base_image(WT_SESSION_IMPL *session, WT_REF *ref,
+  WT_CELL_UNPACK_DELTA_INT **delta, size_t delta_entries, WT_REF ***refsp, size_t *ref_entriesp,
+  size_t *incr)
+{
+    WT_CELL_UNPACK_ADDR *base, *base_key, *base_val;
+    WT_DECL_RET;
+    WT_ITEM base_key_buf, delta_key_buf;
+    WT_PAGE *page;
+    WT_REF **refs;
+    size_t i, j, k, base_entries, estimated_entries, final_entries;
+    int cmp;
+
+    final_entries = i = j = k = 0;
+    page = ref->page;
+    base_entries = (size_t)page->dsk->u.entries;
+
+    WT_ASSERT(session, base_entries != 0 && delta_entries != 0);
+
+    /* Unpack all entries from the base image into an array. */
+    WT_ERR(__wt_calloc_def(session, base_entries, &base));
+    WT_CELL_FOREACH_ADDR (session, page->dsk, base[k]) {
+        k++;
+    }
+    WT_CELL_FOREACH_END;
+
+    /*
+     * Creates a new reference array containing the finalized refs. The maximum number of entries is
+     * the sum of half the base entries (since keys and values are stored separately) and the delta
+     * entries.
+     */
+    estimated_entries = (base_entries / 2) + delta_entries + 1;
```

Comment:
May be improve the comments to say the entries in the base image is the total of key and value

Reply:
Updated [90e2431](https://github.com/wiredtiger/wiredtiger/pull/11627/commits/90e2431cc96edec6ab11a4e7298c5981ce5271a4)